src/NHibernate.Test/FilterTest/DynamicFilterTest.cs

```diff
@@ -354,6 +354,35 @@ public void InStyleFilterParameter()
 			}
 		}
 
+		[Test]
+		public void InStyleFilterParameterWithHashSet()
+		{
+			//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+			// one-to-many loading tests
+			//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+			log.Info("Starting one-to-many collection loader filter tests with HashSet.");
+			using (var session = OpenSession())
+			{
+				Assert.Multiple(
+					() =>
+					{
+						session.EnableFilter("regionlist")
+						       .SetParameterList("regions", new HashSet<string> { "LA", "APAC" });
+
+						log.Debug("Performing query of Salespersons");
+						var salespersons = session.CreateQuery("from Salesperson").List();
+						Assert.That(salespersons.Count, Is.EqualTo(1), "Incorrect salesperson count");
+
+						session.EnableFilter("guidlist")
+						       .SetParameterList("guids", new HashSet<Guid> { testData.Product1Guid, testData.Product2Guid });
+
+						log.Debug("Performing query of Products");
+						var products = session.CreateQuery("from Product").List();
```

Comment:
Before the fix, this case throws an invalid cast exception, which seems dependent on the dataprovider.